File: evd_ros_backend/evd_ros_core/src/evd_script/program/machine_operations/machine_wait.py
'''
Stalls robot's execution until the machine signals that it has completed its
process.

This behavior is required on implemetation of a machine lest programs assume
a static timing (bad programming).
'''

from .machine_primitive import MachinePrimitive


class MachineWait(MachinePrimitive):

    '''
    Data structure methods
    '''

    @classmethod
    def type_string(cls, trailing_delim=True):
        return 'machine-wait' + ('.' if trailing_delim else '')

    @classmethod
    def full_type_string(cls):
        return MachinePrimitive.full_type_string() + cls.type_string()

    def __init__(self, machineUuid=None, type='', name='', uuid=None, parent=None, append_type=True):
        super(MachineWait,self).__init__(
            machineUuid=machineUuid,
            type=MachineWait.type_string() + type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type)

    '''
    Execution methods
    '''

    def symbolic_execution(self, hooks):
        pass

    def realtime_execution(self, hooks):
        pass

File: evd_authoring_app/src/model/evdScript/program/machineOperations/machineWait.js
import { MachinePrimitive } from './machinePrimitive';

export class MachineWait extends MachinePrimitive {

    /*
    * Data structure methods
    */

    static typeString() {
        return 'machine-wait';
    }

    static fullTypeString() {
        return MachinePrimitive.fullTypeString() + MachineWait.typeString();
    }

    constructor(machineUuid=null, type='', name='', uuid=null, parent=null, appendType=true) {
        super(
            machineUuid,
            (appendType) ? 'machine-wait'+type : type,
            name,
            uuid,
            parent,
            appendType
        );
    }

    static fromDict(dct) {
        return new MachineWait(
            dct.machine_uuid,
            dct.type,
            dct.name,
            dct.uuid,
            null,
            false
        );
    }

    static BlocklyToolbox() {
        return { type: 'machine_wait' };
    }

    static BlocklyBlock() {
        return { key: 'machine_wait', data: {
            init: function() {
                this.appendDummyInput()
                    .appendField("Machine Wait");
                this.appendValueInput("machine")
                    .setCheck("machine")
                    .appendField("Machine");
                this.setPreviousStatement(true, null);
                this.setNextStatement(true, null);
                this.setColour(120);
                this.setTooltip("machine-wait");
                this.setHelpUrl("machine-wait");
            }
        }};
    }
}